server: Add RSA precomputation and multiprime benchmarks

joshlf opened the pull request:
Most of the time, RSA private keys should already include cached precomputed values. This PR adds a check that precomputes them in case they haven't been already. It's mostly a hedge against client code forgetting to precompute, which should be uncommon since most of the standard library precomputes under the hood when generating private keys (either from scratch or when parsing files).

lziest commented:
lgtm

joshlf commented:
I'll ask for an "official" review then :)

tmthrgd commented:
`(*rsa.PrivateKey).Precompute()` isn't thread-safe, so this pull request will cause a race condition. It could also lead to pretty severe signing failures as the `decrypt(io.Reader, *PrivateKey, *big.Int) (*big.Int, error)` function in `crypto/rsa` will see partially calculated values. (This race condition would probably linger for a long time causing occasional obscure failures).

*(I don't actually work for Cloudflare but I find this product really interesting).*

joshlf commented:
That is a very good point!

I don't see a good way to do this that both preserves performance, is correct, and doesn't add a lot of complexity. Given the marginal benefit, I think I'll kill the precomputation and just keep the benchmarks.

Thanks @tmthrgd !

joshlf commented:
OK I've removed the precomputation code, added a comment to `Keystore.Get` suggesting precomputation, and added benchmarks for precomputation and, while I was at it, multi-prime RSA.

@lziest or @Bren2010 , it's ready for another review.
